File: frontend/lib/api/secure-india.ts
import {apiClient, type ApiRequestOptions} from "@/lib/api/client";

export type SecureIndiaSource = {dataset_id: string; methodology: string; period_end: string; published_at: string; source_label: string; source_type: "SYNTHETIC"; version: string};
export type SecureIndiaRegion = {bucket: number; city: string; count: number; id: string; state: string; trend_percent: number; value: number; x: number; y: number; zone: string};
export type SecureIndiaLegendBin = {index: number; max: number | null; min: number};
export type SecureIndiaSummary = {
  available_cities: string[];
  available_states: string[];
  filters: {city: string; crime_type: string; period: "7d" | "30d" | "1y"; state: string; view: "count" | "per_lakh"};
  hot_crimes: Array<{count: number; id: string; resource_slug: string; share_percent: number}>;
  hot_zones: SecureIndiaRegion[];
  legend: SecureIndiaLegendBin[];
  map_regions: SecureIndiaRegion[];
  metrics: Array<{id: string; value: number}>;
  rankings: SecureIndiaRegion[];
  source: SecureIndiaSource;
};

export const secureIndiaApi = {
  summary: (query: URLSearchParams, options?: ApiRequestOptions) => apiClient.get<SecureIndiaSummary>("/secure-india/summary?" + query.toString(), options)
};

File: backend/app/schemas/secure_india.py
from typing import Literal

from pydantic import BaseModel


class SecureIndiaSource(BaseModel):
    dataset_id: str
    version: str
    source_type: Literal["SYNTHETIC"]
    source_label: str
    published_at: str
    period_end: str
    methodology: str


class SecureIndiaFilters(BaseModel):
    crime_type: str
    state: str
    city: str
    period: Literal["7d", "30d", "1y"]
    view: Literal["count", "per_lakh"]


class SecureIndiaMetric(BaseModel):
    id: str
    value: float


class SecureIndiaRegion(BaseModel):
    id: str
    city: str
    state: str
    zone: str
    x: float
    y: float
    count: int
    value: float
    trend_percent: int
    bucket: int


class SecureIndiaCrime(BaseModel):
    id: str
    count: int
    share_percent: float
    resource_slug: str


class SecureIndiaLegendBin(BaseModel):
    index: int
    min: float
    max: float | None


class SecureIndiaSummary(BaseModel):
    source: SecureIndiaSource
    filters: SecureIndiaFilters
    available_states: list[str]
    available_cities: list[str]
    metrics: list[SecureIndiaMetric]
    legend: list[SecureIndiaLegendBin]
    map_regions: list[SecureIndiaRegion]
    rankings: list[SecureIndiaRegion]
    hot_zones: list[SecureIndiaRegion]
    hot_crimes: list[SecureIndiaCrime]

